Production/src/Selectors.cc

```diff
@@ -177,5 +179,123 @@ TauJetSelector::Result TauJetTag::Select(const edm::Event& event, const std::deq
+                                     const reco::Vertex& primaryVertex,
+                                     const pat::TriggerObjectStandAloneCollection& triggerObjects,
+                                     const edm::TriggerResults& triggerResults, float rho)
+{
+    const pat::Muon *ref_muon = nullptr;
+    const pat::Electron *ref_electron = nullptr;
+    for(const pat::Muon& muon : muons) { // Look for a muon tag first
+        if(!(muon.pt() > 26 && std::abs(muon.eta()) < 2.1 && muon.isMediumMuon() && PFRelIsolation(muon) < 0.1
+                && std::abs(muon.muonBestTrack()->dz(primaryVertex.position())) < 0.2
+                && std::abs(muon.muonBestTrack()->dxy(primaryVertex.position())) < 0.045))
+            continue;
+        if(!ref_muon || PFRelIsolation(*ref_muon) > PFRelIsolation(muon) || (PFRelIsolation(*ref_muon) == PFRelIsolation(muon) && ref_muon->pt() < muon.pt())){
+            ref_muon = &muon;
+        }
+    }
+    if(!ref_muon) {
+    	for(const pat::Electron& electron : electrons) { // Start looking for an ele tag if no muon was found
+            if(!(electron.pt() > 35 && std::abs(electron.eta()) < 2.3 && electron.electronID("mvaEleID-Fall17-noIso-V2-wp90") > 0.5f
+                     && std::abs(electron.bestTrack()->dz(primaryVertex.position())) < 0.2
+                     && std::abs(electron.bestTrack()->dxy(primaryVertex.position())) < 0.045))
+            	 continue;
+            if(!ref_electron || ref_electron->pt() < electron.pt()){
+                ref_electron = &electron;
+            }
+        }
+    }
+    //
+    auto tagObject = std::make_shared<TagObject>();
+    std::vector<const TauJet*> selectedTauJets;
+    if(ref_muon) { // Fill tag object with muon if muon found
+        tagObject->kind = tau_analysis::selectors::TagObject::Kind::Muon; //new tag object member 
+        tagObject->p4 = ref_muon->polarP4();
+        tagObject->charge = ref_muon->charge();
+        tagObject->id = unsigned(ref_muon->isLooseMuon()) * 1 + unsigned(ref_muon->isMediumMuon()) * 2
+                          + unsigned(ref_muon->isTightMuon(primaryVertex)) * 4;
+        tagObject->isolation = PFRelIsolation(*ref_muon);
+        tagObject->has_extramuon = hasExtraMuon(muons, ref_muon, primaryVertex);
+        tagObject->has_extraelectron = hasExtraElectron(electrons, rho);
+        tagObject->has_dimuon = hasExtraDimuon(muons, primaryVertex);
+        const TauJet* selected = nullptr;
+        for(const TauJet& tauJet : tauJets) { //Look for highest pT jet
+            if(!tauJet.jet) continue;
+	    const pat::Jet& jet = *tauJet.jet;
+            if(!(std::abs(jet.eta()) < 2.8 && reco::deltaR(jet.polarP4(), ref_muon->polarP4()) > 0.8))
+		    continue;
+	    if(!selected || selected->jet->pt() < jet.pt()) {
+		    selected = &tauJet;
+		    selectedTauJets = { selected };
+            }
+	}
+	return Result(selectedTauJets, tagObject);
```

Comment:
Unify code to select tauJet for ref_muon, ref_electron and "ref_MET" cases to avoid duplications.